hikari/impl/event_manager_base.py

```diff
@@ -50,46 +58,161 @@
     ConsumerT = typing.Callable[
         [gateway_shard.GatewayShard, data_binding.JSONObject], typing.Coroutine[typing.Any, typing.Any, None]
     ]
-    ListenerMapT = typing.MutableMapping[
+    ListenerMapT = typing.Dict[
         typing.Type[event_manager.EventT_co],
-        typing.MutableSequence[event_manager.CallbackT[event_manager.EventT_co]],
+        typing.List[event_manager.CallbackT[event_manager.EventT_co]],
     ]
     WaiterT = typing.Tuple[event_manager.PredicateT[event_manager.EventT_co], asyncio.Future[event_manager.EventT_co]]
-    WaiterMapT = typing.MutableMapping[
-        typing.Type[event_manager.EventT_co], typing.MutableSet[WaiterT[event_manager.EventT_co]]
+    WaiterMapT = typing.Dict[typing.Type[event_manager.EventT_co], typing.Set[WaiterT[event_manager.EventT_co]]]
+
+    EventManagerBaseT = typing.TypeVar("EventManagerBaseT", bound="EventManagerBase")
+    UnboundMethodT = typing.Callable[
+        [EventManagerBaseT, gateway_shard.GatewayShard, data_binding.JSONObject],
+        typing.Coroutine[typing.Any, typing.Any, None],
     ]
 
 
+@typing.runtime_checkable
+class _FilteredMethodT(fast_protocol.FastProtocolChecking, typing.Protocol):
+    async def __call__(self, shard: gateway_shard.GatewayShard, payload: data_binding.JSONObject, /) -> None:
+        raise NotImplementedError
+
+    @property
+    def __cache_components__(self) -> config.CacheComponents:
+        raise NotImplementedError
+
+    @property
+    def __event_types__(self) -> typing.Sequence[typing.Type[base_events.Event]]:
+        raise NotImplementedError
+
+
 def _default_predicate(_: event_manager.EventT_inv) -> bool:
     return True
 
 
+def filtered(
+    event_types: typing.Union[typing.Type[base_events.Event], typing.Sequence[typing.Type[base_events.Event]]],
+    cache_components: config.CacheComponents = config.CacheComponents.NONE,
+    /,
+) -> typing.Callable[[UnboundMethodT[EventManagerBaseT]], UnboundMethodT[EventManagerBaseT]]:
```

Comment:
Could this be changed to mark consumers instead?

Reply:
That doesn't really bring any benefit as a change (plus it seems a bit odd to enforce external filtering when not all event consumers need to be externally filtered) (and fwiw that'd be a breaking change for anybody who's overriding methods on this class through inheritance)